federationsender/queue/destinationqueue.go

```diff
@@ -128,57 +160,86 @@ func (oq *destinationQueue) backgroundSend() {
 	}
 	defer oq.running.Store(false)
 
+	// Since the PDU buffer can be rehydrated from the database, we
+	// dont need to keep it in memory after the worker stops running.
+	// transactionID and transactionCount are enough to help queue up
+	// additional event in the same transaction.
+	defer oq.cleanPendingPDUs()
+
 	for {
+		// For now we don't know the next transaction ID. Set it to an
+		// empty one. The next step will populate it if we have pending
+		// PDUs in the database. Otherwise we'll generate one later on,
+		// e.g. in response to EDUs.
+		transactionID := gomatrixserverlib.TransactionID("")
+
+		// Retrieve the events in the next transaction. Note that this
+		// does *not* necessarily mean we will fill the buffer - the
+		// important thing here is that we still continue to send
+		// transactions in order.
+		hydrate := func() {
+			txid, pdus, err := oq.db.GetNextTransactionPDUs(
+				context.TODO(), // context
+				oq.destination, // server name
+				maxPDUsPerTransaction-len(oq.pendingPDUs), // how many events to retrieve
+			)
+			if err != nil {
+				log.WithError(err).Errorf("failed to get next transaction PDUs for server %q", oq.destination)
+				return
+			}
+			transactionID = txid
+			oq.pendingPDUs = append(oq.pendingPDUs, pdus...)
+		}
+
+		// If we haven't reached the PDU limit yet then rehydrate the
+		// PDU queue from the database.
+		if len(oq.pendingPDUs) < maxPDUsPerTransaction {
+			hydrate()
```

Comment:
I'd prefer if we removed `oq.pendingPDUs` completely and just had function-level vars now. It's weird to have all these side-effects.